File: test/.work/robust_optional_matrix/ts/sb/struct_matrix.ts
import * as _ from "./_"
import * as Enum from "./enum"

export interface Matrix extends _.Serializable, _.Deserializable {
    single: _.Child | null;
    many: _.Child[];
    kind: Enum.Kind;
    kinds: Enum.Kind[];
}

export const newMatrix = (): Matrix => {
    const s = {
        single: null,
        many: [],
        kind: 0,
        kinds: [],
    } as any as Matrix;
    s.set = (buf: _.Buffer) => setMatrix(buf, s);
    s.get = (buf: _.Buffer) => {
        const [res, err] = getMatrix(buf);
        if (err === null) Object.assign(s, res);
        return err;
    };
    return s;
}

export const eqMatrix = (a: Matrix, b: Matrix): boolean => {
    if (a === b) return true;
    if (a === null || b === null) return false;
    if (!_.eqChild(a.single, b.single)) return false;
    if (!_.eqChildList(a.many, b.many)) return false;
    if (a.kind !== b.kind) return false;
    if (!_.eqU8List(a.kinds as any, b.kinds as any)) return false;
    return true;
}

export const getMatrix = (buf: _.Buffer): [Matrix, Error | null] => {
    const s = newMatrix();
    const bitmaskSize = Math.ceil(4 / 8);
    const [bits, err] = buf.read(bitmaskSize);
    if (err !== null) return [s, err];
    if (_.GetBit(bits, 0)) {
        const [v, err] = _.getChild(buf);
        if (err !== null) return [s, err];
        s.single = v;
    }
    if (_.GetBit(bits, 1)) {
        const [v, err] = _.getChildList(buf);
        if (err !== null) return [s, err];
        s.many = v;
    }
    if (_.GetBit(bits, 2)) {
        const [v, err] = _.getU8(buf);
        if (err !== null) return [s, err];
        s.kind = v as any;
        if (!_.IsKind(s.kind as any)) return [s, new Error("get Matrix Kind: invalid enum value")];
    }
    if (_.GetBit(bits, 3)) {
        const [v, err] = _.getU8List(buf);
        if (err !== null) return [s, err];
        s.kinds = v as any;
        if (!_.IsKindList(s.kinds as any)) return [s, new Error("get Matrix Kinds: invalid enum value")];
    }
    return [s, null];
}

export const setMatrix = (buf: _.Buffer, s: Matrix): Error | null => {
    if (s === null || s === undefined) return new Error(`set Matrix: value is null or undefined`);
    const bits = new Uint8Array(Math.ceil(4 / 8));
    const body = new _.Buffer();
    if (s.single !== null && s.single !== undefined) {
        const err = _.setChild(body, s.single);
        if (err !== null) return err;
        _.SetBit(bits, 0, true);
    }
    if (s.many && s.many.length > 0) {
        const err = _.setChildList(body, s.many);
        if (err !== null) return err;
        _.SetBit(bits, 1, true);
    }
    if ((s.kind as any) !== 0) {
        if (!_.IsKind(s.kind as any)) return new Error("set Matrix Kind: invalid enum value");
        const err = _.setU8(body, s.kind as any);
        if (err !== null) return err;
        _.SetBit(bits, 2, true);
    }
    if (s.kinds && s.kinds.length > 0) {
        if (!_.IsKindList(s.kinds as any)) return new Error("set Matrix Kinds: invalid enum value");
        const err = _.setU8List(body, s.kinds as any);
        if (err !== null) return err;
        _.SetBit(bits, 3, true);
    }

    const errBits = buf.write(bits);
    if (errBits !== null) return errBits;
    return buf.write(body.bytes);
}

export const getMatrixList = (buf: _.Buffer): [Matrix[], Error | null] => _.getList(buf, getMatrix);
export const setMatrixList = (buf: _.Buffer, v: Matrix[]): Error | null => _.setList(buf, v, setMatrix);
export const eqMatrixList = (a: Matrix[], b: Matrix[]): boolean => _.eqList(a, b, eqMatrix);

File: test/.work/robust_optional_matrix/go/sb/struct_matrix.go
package sb

import (
	"bytes"
	"fmt"
	"math"
	"slices"
)

type Matrix struct {
	Single *Child  
	Many []*Child  
	Kind Kind  
	Kinds []Kind  
}

func GetMatrix(buf *bytes.Buffer, s *Matrix) error {
	if s == nil { return nil }
	bitSize := int(math.Ceil(float64(4) / 8.0))
	if buf.Len() < bitSize { return fmt.Errorf("GetMatrix bitmask: %d - %d", buf.Len(), bitSize) }
	bits := buf.Next(bitSize)
	if GetBit(bits, uint8(0)) {
		val, err := ReadChild(buf)
		if err != nil { return fmt.Errorf("GetMatrix Single: %w", err) }
		s.Single = val
	}
	if GetBit(bits, uint8(1)) {
		val, err := GetChildList(buf)
		if err != nil { return fmt.Errorf("GetMatrix Many: %w", err) }
		s.Many = val
	}
	if GetBit(bits, uint8(2)) {
		val, err := GetKind(buf)
		if err != nil { return fmt.Errorf("GetMatrix Kind: %w", err) }
		s.Kind = val
		if !IsKind(s.Kind) { return fmt.Errorf("GetMatrix Kind: 非法枚举值: %d", s.Kind) }
	}
	if GetBit(bits, uint8(3)) {
		val, err := GetKindList(buf)
		if err != nil { return fmt.Errorf("GetMatrix Kinds: %w", err) }
		s.Kinds = val
	}
	if err := ValidateMatrix(s); err != nil { return fmt.Errorf("ValidateMatrix: %w", err) }
	return nil
}

func SetMatrix(buf *bytes.Buffer, s *Matrix) error {
	if s == nil { return fmt.Errorf("SetMatrix: nil value") }
	if err := ValidateMatrix(s); err != nil { return fmt.Errorf("ValidateMatrix: %w", err) }
	bits := make([]byte, uint8(math.Ceil(float64(4)/8.0)))
	body := bytes.NewBuffer(nil)
	if s.Single != nil {
		if err := SetChild(body, s.Single); err != nil { return fmt.Errorf("SetMatrix Single: %w", err) }
		SetBit(bits, uint8(0), true)
	}
	if len(s.Many) > 0 {
		if err := SetChildList(body, (ChildList)(s.Many)); err != nil { return fmt.Errorf("SetMatrix Many: %w", err) }
		SetBit(bits, uint8(1), true)
	}
	if s.Kind != 0 {
		if err := SetKind(body, s.Kind); err != nil { return fmt.Errorf("SetMatrix Kind: %w", err) }
		SetBit(bits, uint8(2), true)
	}
	if len(s.Kinds) > 0 {
		if err := SetKindList(body, s.Kinds); err != nil { return fmt.Errorf("SetMatrix Kinds: %w", err) }
		SetBit(bits, uint8(3), true)
	}

	if _, err := buf.Write(bits); err != nil { return fmt.Errorf("SetMatrix write bitmask: %w", err) }
	_, err := body.WriteTo(buf); return err
}

func ValidateMatrix(s *Matrix) error {
	if s == nil { return nil }
	if s.Single != nil {
		if err := ValidateChild(s.Single); err != nil { return fmt.Errorf("Single: %w", err) }
	}
	if err := ValidateChildList(s.Many); err != nil { return fmt.Errorf("Many: %w", err) }
	if s.Kind != 0 && !IsKind(s.Kind) { return fmt.Errorf("Kind 非法枚举值: %d", s.Kind) }
	for i, item := range s.Kinds {
		if !IsKind(item) { return fmt.Errorf("Kinds[%d] 非法枚举值: %d", i, item) }
	}
	return nil
}

func EqMatrix(a, b *Matrix) bool {
	if a == b { return true }
	if a == nil || b == nil { return false }
	if !EqChild(a.Single, b.Single) { return false }
	if !EqChildList((ChildList)(a.Many), (ChildList)(b.Many)) { return false }
	if a.Kind != b.Kind { return false }
	if !EqKindList(a.Kinds, b.Kinds) { return false }
	return true
}

// Standalone functions
func ReadMatrix(buf *bytes.Buffer) (*Matrix, error) {
	s := new(Matrix)
	return s, GetMatrix(buf, s)
}

type MatrixList []*Matrix
func GetMatrixList(buf *bytes.Buffer) (MatrixList, error) { return getList[*Matrix, MatrixList](buf, ReadMatrix) }
func SetMatrixList(buf *bytes.Buffer, v MatrixList) error { return setList(buf, v, SetMatrix) }
func ValidateMatrixList(v MatrixList) error {
	for i, item := range v {
		if item == nil { return fmt.Errorf("MatrixList[%d]: nil item", i) }
		if err := ValidateMatrix(item); err != nil { return fmt.Errorf("MatrixList[%d]: %w", i, err) }
	}
	return nil
}
func EqMatrixList(a, b MatrixList) bool { return slices.EqualFunc(a, b, EqMatrix) }
